Add ability to navigate to resources page from build page

chenbh opened the pull request:
## What does this PR accomplish?
Bug Fix | **Feature** | Documentation

closes # .

![Screen Shot 2021-03-12 at 5 36 40 PM](https://user-images.githubusercontent.com/17577073/111006056-0c349500-835a-11eb-9397-ce3bdda941e2.png)

## Changes proposed by this PR:

- add ability to open the Resource page to a specific version. If the version exists, then it will be displayed at the top of the page and is automatically expanded
- have the get/put steps in the Build page make use of this

## Notes to reviewer:

## Release Note

UI: clicking on the version text for a get/put step in the Build page will now navigate directly to the Resource page with the corresponding version expanded

## Contributor Checklist
- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [ ] ~Updated [Documentation]~
- [x] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

Commit:
B: navigate directly to a resource version via url

- added version field to resource model
- modify the route to resource page to allow passing in a version
- when the resource page is loaded with a version, it'll attempt to
figure out the id of that version and navigate to a page with that id
- if the resource page is loaded with a version and fetched the list of
versioned resources, any resource that matches the version will be auto
expanded

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
S: add tests for resource page with version

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
B: link to resource page from build page

- string a JobBuildIdentifier through the steptree so we can get a
PipelineIdentifier to figure out href
- add hover tooltip

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
S: add tests for get/put step version hrefs

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
S: fix benchmarks

Signed-off-by: Bohan Chen <bochen@pivotal.io>

aoldershaw commented:
This is a very nice feature to have - frequently I've found myself copy/pasting the version I care about and pasting it into the browser's "find" feature, which is slow + annoying + tedious if the resource isn't on the first page.

A couple notes:
* Based on how we add arrows to tooltips (they're positioned centered on the element, regardless of where the tooltip actually is), the tooltips look weird when the version table is wide since the arrow is just hovering:
    <img width="649" alt="Screen Shot 2021-03-15 at 11 48 13 AM" src="https://user-images.githubusercontent.com/26583442/111183825-b55fd300-8586-11eb-9009-6ef130360f7e.png">
    * We can either try to center the tooltip box (which requires hard-coding its width), or just get rid of the arrow (which is the approach I took with other potentially wide elements)
* We shouldn't display this tooltip for anonymous `image_resources` - currently, it tries to take you to the resource page for a resource named `image`, since that's the step name
* Generally, we shouldn't be using the step name to determine which resource page to go to - we should use the `resource` field
  * e.g. if you have the following `get` step:
  ```yaml
  - get: foo
    resource: bar
  ```
  the link takes you to `.../resources/foo`, rather than `.../resources/bar`. I think we can also use this to determine if the resource is an `image_resource`, since `resource` should be empty

Note: we already expose the `resource` on the public plan: https://github.com/concourse/concourse/blob/28153d0cb8dd6f670d09a2001bd7b27d74ddf217/atc/public_plan.go#L183

...so just need to modify the JSON decoding to decode that in Elm:
https://github.com/concourse/concourse/blob/28153d0cb8dd6f670d09a2001bd7b27d74ddf217/web/elm/src/Concourse.elm#L665-L669

Commit:
B: use resource instead of step names for url link

- parse the `resource` field from the public build plan
- use `resource` instead of `name` field to link to resource
- disable linking on anonymous get steps

Signed-off-by: Bohan Chen <bochen@pivotal.io>

chenbh commented:
Good call, I didn't even think about anonymous gets

aoldershaw commented on web/elm/src/Build/StepTree/StepTree.elm:
```diff
@@ -1274,6 +1340,20 @@ tooltip model { hovered } =
+                        [ Html.text "view in resources page" ]
+                , attachPosition =
+                    { direction = Tooltip.Top
+                    , alignment = Tooltip.End
```
Could we actually to `Tooltip.Start`? Might just be me, but when I was trying to use this feature, I tended to put my cursor closer to the start of the version table (near where the keys are). When the version is particularly wide (like  with the `registry-image`), the tooltip appears pretty far away

chenbh replied:
I briefly tried this before, but there was something weird about how it calculated the tooltip width and made it longer than necessary. I also liked the way the alignment was sorta consistent in style with the image fetching and step initialization timer tooltips.

Maybe I'll just bite the bullet and figure out the values to make it `Tooltip.Middle`

chenbh replied:
Ended up doing `Tooltip.Middle`. 

> but there was something weird about how it calculated the tooltip width and made it longer than necessary.

This happens if the version text is short enough, the tooltip overflows the side of the screen into a newline, but the tooltip box doesn't take that into consideration and is longer than it need to be.
![Screen Shot 2021-03-23 at 11 30 21 AM](https://user-images.githubusercontent.com/17577073/112172665-4825f080-8bcb-11eb-9863-75c3b017618c.png)

But I'm going to say that versions text shorter than 9 characters (I think that's around the limit) are rare enough and we don't have to worry about it.

aoldershaw commented on web/elm/src/Routes.elm:
```diff
@@ -188,9 +190,34 @@ resource =
+        split s =
+            case String.split ":" s of
+                [ k, v ] ->
+                    Just ( k, v )
```
We'll need to rethink this a bit, as it doesn't work if the key/value contains a `:` (which happens a lot, actually, since the `digest` of the `registry-image` resource has a `:`) - if I click on the version of a resource using `registry-image`, it just strips out the `digest` from the URL and only keeps the `tag`.

The easiest solution would be to just encode a full JSON object and throw that in a single `filter` query param.

Another fairly easy solution is for `Concourse.versionQuery` to `URL.percentEncode` the key and value separately, and `URL.percentDecode` them separately here as well. That way, they end up getting doubly-encoded. A similar approach is to base64 encode them, but that makes the query string a little opaque.

If you want to avoid an ugly query string (because of all the special characters that need escaping in the other approaches), you could do something like I did with instance vars (`vars.key1="value1"&vars.key2="value2"`) - the serialization/parsing is all in https://github.com/concourse/concourse/blob/master/web/elm/src/DotNotation.elm. Since it's just a `Dict String String` it's even simpler - you don't need to worry about recursing through JSON objects or quoting strings with special characters. However, this approach requires a bit of a hack since the query parameter names are dynamic:
https://github.com/concourse/concourse/blob/ecda93e38208d0971e280076f9fbe8864d8e90dc/web/elm/src/Routes.elm#L521-L532

I personally think the JSON approach is the easiest and probably the best - I tried to make the instance vars query params more friendly since they appear all the time, but in this case you probably won't need to see it that often.

chenbh replied:
I originally had it in the JSON approach, but I took that out to make it more similar to the atc api logic. My thought was that if it fails here, then it'll also fail when the api server tries to decode it. What do you think of just taking the head of the list as the key and concatenate the rest into a string for the value?

https://github.com/concourse/concourse/blob/fc2fda3221c19f7ba39b2e9b96fa492463129e34/atc/api/resourceserver/versionserver/list.go#L34-L41

aoldershaw replied:
Huh, wasn't aware of that! seems a bit problematic in general that the API server doesn't allow for keys that have a `:`, but oh well. I think your suggestion of taking the head as the key and the tail (joined by `:`) as the value is reasonable, since that'd match the behaviour of `SplitN(field, ":", 2)`

Commit:
web: allow multiple : in version parsing from url

also moved the version tooltip to align middle

Signed-off-by: Bohan Chen <bochen@pivotal.io>

aoldershaw reviewed (approved):
LGTM!

xtremerui commented:
if you approve fast enough would you beat the checks to merge it lol

chenbh commented:
> if you approve fast enough would you beat the checks to merge it lol

Not if the checks are marked as `Required`
